File: android/src/main/kotlin/com/flutternative3d/flutter_native_3d/Native3DPlatformView.kt
package com.flutternative3d.flutter_native_3d

import android.content.Context
import android.view.View
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView

/**
 * Flutter platform view that bridges method channel calls to [SceneManager].
 *
 * This class handles only Flutter communication -- no rendering logic.
 * All 3D operations are delegated to SceneManager.
 *
 * Channel: `flutter_native_3d/scene_$viewId`
 */
class Native3DPlatformView(
    context: Context,
    viewId: Int,
    creationParams: Map<String, Any>,
    messenger: BinaryMessenger
) : PlatformView, MethodChannel.MethodCallHandler {

    private val channel = MethodChannel(messenger, "flutter_native_3d/scene_$viewId")
    private val sceneManager = SceneManager(context)
    private var disposed = false

    init {
        channel.setMethodCallHandler(this)
        applyCreationParams(creationParams)

        // Signal that the native scene is ready for commands
        android.os.Handler(android.os.Looper.getMainLooper()).post {
            if (!disposed) {
                channel.invokeMethod("onEvent", mapOf("type" to "sceneReady"))
            }
        }
    }

    override fun getView(): View = sceneManager.view

    override fun dispose() {
        if (disposed) return
        disposed = true
        channel.setMethodCallHandler(null)
        sceneManager.dispose()
    }

    // -------------------------------------------------------------------------
    // Creation params (applied once at native view init)
    // -------------------------------------------------------------------------

    private fun applyCreationParams(params: Map<String, Any>) {
        (params["backgroundColor"] as? Int)?.let {
            sceneManager.setBackgroundColor(it)
        }
        (params["lighting"] as? String)?.let {
            sceneManager.setLighting(it)
        }
        (params["fitMode"] as? String)?.let {
            sceneManager.setFitMode(it)
        }
        (params["gesturesEnabled"] as? Boolean)?.let {
            sceneManager.setGesturesEnabled(it)
        }
        (params["autoRotate"] as? Boolean)?.let {
            sceneManager.setAutoRotate(it)
        }
        @Suppress("UNCHECKED_CAST")
        (params["initialCameraOrbit"] as? Map<String, Any>)?.let { orbit ->
            val theta = (orbit["theta"] as? Number)?.toFloat() ?: 0f
            val phi = (orbit["phi"] as? Number)?.toFloat() ?: 20f
            val radius = (orbit["radius"] as? Number)?.toFloat() ?: 3f
            sceneManager.setCameraOrbit(theta, phi, radius)
        }

        // NOTE: We intentionally do NOT load the model here from creation params.
        // The Dart side sends loadModel via the method channel so it can receive
        // the Future<ModelInfo> result. Loading here too would cause a double load.
    }

    // -------------------------------------------------------------------------
    // Method channel dispatch
    // -------------------------------------------------------------------------

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        if (disposed) {
            result.error("DISPOSED", "Native view has been disposed", null)
            return
        }

        when (call.method) {
            "loadModel" -> handleLoadModel(call, result)
            "resetCamera" -> {
                sceneManager.resetCamera()
                result.success(null)
            }
            "setCameraOrbit" -> {
                val theta = call.argument<Number>("theta")?.toFloat() ?: 0f
                val phi = call.argument<Number>("phi")?.toFloat() ?: 20f
                val radius = call.argument<Number>("radius")?.toFloat() ?: 3f
                sceneManager.setCameraOrbit(theta, phi, radius)
                result.success(null)
            }
            "getAnimationNames" -> {
                result.success(sceneManager.getAnimationNames())
            }
            "playAnimation" -> {
                val name = call.argument<String>("name")
                if (name == null) {
                    result.error("INVALID_ARGS", "Expected 'name' argument", null)
                    return
                }
                val loop = call.argument<Boolean>("loop") ?: true
                val error = sceneManager.playAnimation(name, loop) { completedName ->
                    if (!disposed) {
                        channel.invokeMethod("onEvent", mapOf(
                            "type" to "animationCompleted",
                            "name" to completedName,
                        ))
                    }
                }
                if (error != null) {
                    result.error("ANIMATION_ERROR", error, null)
                } else {
                    result.success(null)
                }
            }
            "playAnimationByIndex" -> {
                val index = call.argument<Int>("index")
                if (index == null) {
                    result.error("INVALID_ARGS", "Expected 'index' argument", null)
                    return
                }
                val loop = call.argument<Boolean>("loop") ?: true
                val error = sceneManager.playAnimationByIndex(index, loop) { completedName ->
                    if (!disposed) {
                        channel.invokeMethod("onEvent", mapOf(
                            "type" to "animationCompleted",
                            "name" to completedName,
                        ))
                    }
                }
                if (error != null) {
                    result.error("ANIMATION_ERROR", error, null)
                } else {
                    result.success(null)
                }
            }
            "pauseAnimation" -> {
                sceneManager.pauseAnimation()
                result.success(null)
            }
            "stopAnimation" -> {
                sceneManager.stopAnimation()
                result.success(null)
            }
            "setBackgroundColor" -> {
                val color = call.argument<Int>("color")
                if (color == null) {
                    result.error("INVALID_ARGS", "Expected 'color' argument", null)
                    return
                }
                sceneManager.setBackgroundColor(color)
                result.success(null)
            }
            "setLighting" -> {
                val preset = call.argument<String>("preset")
                if (preset == null) {
                    result.error("INVALID_ARGS", "Expected 'preset' argument", null)
                    return
                }
                sceneManager.setLighting(preset)
                result.success(null)
            }
            "setGesturesEnabled" -> {
                val enabled = call.argument<Boolean>("enabled") ?: true
                sceneManager.setGesturesEnabled(enabled)
                result.success(null)
            }
            "setAutoRotate" -> {
                val enabled = call.argument<Boolean>("enabled") ?: false
                sceneManager.setAutoRotate(enabled)
                result.success(null)
            }
            "setFitMode" -> {
                val mode = call.argument<String>("mode") ?: "contain"
                sceneManager.setFitMode(mode)
                result.success(null)
            }
            "dispose" -> {
                dispose()
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    private fun handleLoadModel(call: MethodCall, result: MethodChannel.Result) {
        @Suppress("UNCHECKED_CAST")
        val args = call.arguments as? Map<String, Any>
        if (args == null) {
            result.error("INVALID_ARGS", "Expected map with 'type' and 'path'", null)
            return
        }

        val onProgress: (Float) -> Unit = { progress ->
            if (!disposed) {
                channel.invokeMethod("onEvent", mapOf(
                    "type" to "loadProgress",
                    "progress" to progress.toDouble(),
                ))
            }
        }

        sceneManager.loadModel(args, autoPlay = false, onProgress = onProgress) { loadResult ->
            if (disposed) return@loadModel
            loadResult.fold(
                onSuccess = { info -> result.success(info) },
                onFailure = { error ->
                    result.error(
                        errorCode(error),
                        error.message ?: "Unknown error loading model",
                        error.stackTraceToString()
                    )
                }
            )
        }
    }

    /// Map a native error to a structured error code for the Dart side.
    private fun errorCode(error: Throwable): String {
        val message = error.message?.lowercase() ?: ""
        return when {
            message.contains("asset not found") || message.contains("not found in assets") -> "ASSET_NOT_FOUND"
            message.contains("file not found") -> "FILE_NOT_FOUND"
            message.contains("file not readable") -> "FILE_NOT_FOUND"
            message.contains("http ") || message.contains("download") || message.contains("connect") -> "NETWORK_ERROR"
            message.contains("unsupported file format") || message.contains("unsupported") -> "FORMAT_ERROR"
            else -> "MODEL_LOAD_ERROR"
        }
    }
}

File: ios/Classes/Native3DPlatformView.swift
import Flutter
import UIKit

/// Flutter platform view that bridges method channel calls to SceneManager.
///
/// This class handles only Flutter communication -- no rendering logic.
/// All 3D operations are delegated to SceneManager.
///
/// Channel: `flutter_native_3d/scene_$viewId`
class Native3DPlatformView: NSObject, FlutterPlatformView {
    private let sceneManager: SceneManager
    private let channel: FlutterMethodChannel
    private var disposed = false

    init(frame: CGRect, viewId: Int64, args: [String: Any], messenger: FlutterBinaryMessenger) {
        channel = FlutterMethodChannel(
            name: "flutter_native_3d/scene_\(viewId)",
            binaryMessenger: messenger
        )
        sceneManager = SceneManager(frame: frame)

        super.init()

        channel.setMethodCallHandler(handle)
        applyCreationParams(args)

        // Signal that the native scene is ready for commands
        DispatchQueue.main.async { [weak self] in
            self?.channel.invokeMethod("onEvent", arguments: ["type": "sceneReady"])
        }
    }

    func view() -> UIView {
        return sceneManager.view
    }

    /// Map a native error to a structured error code for the Dart side.
    private static func errorCode(for error: Error) -> String {
        if let loaderError = error as? LoaderError {
            return switch loaderError {
            case .assetNotFound: "ASSET_NOT_FOUND"
            case .fileNotFound, .fileNotReadable: "FILE_NOT_FOUND"
            case .invalidURL, .httpError, .downloadFailed: "NETWORK_ERROR"
            case .unsupportedFormat: "FORMAT_ERROR"
            case .unknownType, .missingField: "MODEL_LOAD_ERROR"
            }
        }
        return "MODEL_LOAD_ERROR"
    }

    // MARK: - Creation Params (applied once at native view init)

    private func applyCreationParams(_ params: [String: Any]) {
        if let bgColor = params["backgroundColor"] as? Int {
            sceneManager.setBackgroundColor(bgColor)
        }
        if let lighting = params["lighting"] as? String {
            sceneManager.setLighting(lighting)
        }
        if let fitMode = params["fitMode"] as? String {
            sceneManager.setFitMode(fitMode)
        }
        if let gesturesEnabled = params["gesturesEnabled"] as? Bool {
            sceneManager.setGesturesEnabled(gesturesEnabled)
        }
        if let autoRotate = params["autoRotate"] as? Bool {
            sceneManager.setAutoRotate(autoRotate)
        }
        if let orbit = params["initialCameraOrbit"] as? [String: Any] {
            let theta = (orbit["theta"] as? NSNumber)?.floatValue ?? 0
            let phi = (orbit["phi"] as? NSNumber)?.floatValue ?? 20
            let radius = (orbit["radius"] as? NSNumber)?.floatValue ?? 3
            sceneManager.setCameraOrbit(theta: theta, phi: phi, radius: radius)
        }

        // NOTE: We intentionally do NOT load the model here from creation params.
        // The Dart side sends loadModel via the method channel so it can receive
        // the Future<ModelInfo> result. Loading here too would cause a double load
        // (creation-params load is discarded, channel load clears and reloads).
    }

    // MARK: - Method Channel Dispatch

    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if disposed {
            result(FlutterError(code: "DISPOSED", message: "Native view has been disposed", details: nil))
            return
        }

        switch call.method {
        case "loadModel":
            handleLoadModel(call, result: result)

        case "resetCamera":
            sceneManager.resetCamera()
            result(nil)

        case "setCameraOrbit":
            let args = call.arguments as? [String: Any] ?? [:]
            let theta = (args["theta"] as? NSNumber)?.floatValue ?? 0
            let phi = (args["phi"] as? NSNumber)?.floatValue ?? 20
            let radius = (args["radius"] as? NSNumber)?.floatValue ?? 3
            sceneManager.setCameraOrbit(theta: theta, phi: phi, radius: radius)
            result(nil)

        case "getAnimationNames":
            result(sceneManager.getAnimationNames())

        case "playAnimation":
            guard let args = call.arguments as? [String: Any],
                  let name = args["name"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Expected 'name' argument", details: nil))
                return
            }
            let loop = args["loop"] as? Bool ?? true
            let onComplete: (String) -> Void = { [weak self] completedName in
                guard let self = self, !self.disposed else { return }
                self.channel.invokeMethod("onEvent", arguments: [
                    "type": "animationCompleted",
                    "name": completedName,
                ])
            }
            if let error = sceneManager.playAnimation(name: name, loop: loop, onComplete: onComplete) {
                result(FlutterError(code: "ANIMATION_ERROR", message: error, details: nil))
            } else {
                result(nil)
            }

        case "playAnimationByIndex":
            guard let args = call.arguments as? [String: Any],
                  let index = args["index"] as? Int else {
                result(FlutterError(code: "INVALID_ARGS", message: "Expected 'index' argument", details: nil))
                return
            }
            let loop = args["loop"] as? Bool ?? true
            let onComplete: (String) -> Void = { [weak self] completedName in
                guard let self = self, !self.disposed else { return }
                self.channel.invokeMethod("onEvent", arguments: [
                    "type": "animationCompleted",
                    "name": completedName,
                ])
            }
            if let error = sceneManager.playAnimationByIndex(index: index, loop: loop, onComplete: onComplete) {
                result(FlutterError(code: "ANIMATION_ERROR", message: error, details: nil))
            } else {
                result(nil)
            }

        case "pauseAnimation":
            sceneManager.pauseAnimation()
            result(nil)

        case "stopAnimation":
            sceneManager.stopAnimation()
            result(nil)

        case "setBackgroundColor":
            guard let args = call.arguments as? [String: Any],
                  let color = args["color"] as? Int else {
                result(FlutterError(code: "INVALID_ARGS", message: "Expected 'color' argument", details: nil))
                return
            }
            sceneManager.setBackgroundColor(color)
            result(nil)

        case "setLighting":
            guard let args = call.arguments as? [String: Any],
                  let preset = args["preset"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Expected 'preset' argument", details: nil))
                return
            }
            sceneManager.setLighting(preset)
            result(nil)

        case "setGesturesEnabled":
            let args = call.arguments as? [String: Any] ?? [:]
            let enabled = args["enabled"] as? Bool ?? true
            sceneManager.setGesturesEnabled(enabled)
            result(nil)

        case "setAutoRotate":
            let args = call.arguments as? [String: Any] ?? [:]
            let enabled = args["enabled"] as? Bool ?? false
            sceneManager.setAutoRotate(enabled)
            result(nil)

        case "setFitMode":
            let args = call.arguments as? [String: Any] ?? [:]
            let mode = args["mode"] as? String ?? "contain"
            sceneManager.setFitMode(mode)
            result(nil)

        case "dispose":
            disposed = true
            channel.setMethodCallHandler(nil)
            sceneManager.dispose()
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func handleLoadModel(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGS", message: "Expected map with 'type' and 'path'", details: nil))
            return
        }

        let onProgress: (Float) -> Void = { [weak self] progress in
            guard let self = self, !self.disposed else { return }
            self.channel.invokeMethod("onEvent", arguments: [
                "type": "loadProgress",
                "progress": Double(progress),
            ])
        }

        sceneManager.loadModel(source: args, autoPlay: false, onProgress: onProgress) { [weak self] loadResult in
            guard let self = self, !self.disposed else { return }
            switch loadResult {
            case .success(let info):
                result(info)
            case .failure(let error):
                result(FlutterError(
                    code: Self.errorCode(for: error),
                    message: error.localizedDescription,
                    details: String(describing: error)
                ))
            }
        }
    }
}
